fabric-biome-api-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheEndBiomeSource.java

```diff
@@ -0,0 +1,81 @@
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.Unique;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import net.minecraft.world.biome.BiomeKeys;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.util.registry.RegistryKey;
+import net.minecraft.world.biome.layer.util.LayerRandomnessSource;
+import net.minecraft.world.biome.source.TheEndBiomeSource;
+
+import net.fabricmc.fabric.impl.biome.InternalBiomeData;
+import net.fabricmc.fabric.impl.biome.SimpleLayerRandomnessSource;
+import net.fabricmc.fabric.impl.biome.WeightedBiomePicker;
+
+@Mixin(TheEndBiomeSource.class)
+public class MixinTheEndBiomeSource {
+	@Shadow
+	@Final
+	private Registry<Biome> biomeRegistry;
+	@Shadow
+	@Final
+	private long seed;
+	@Unique
+	private LayerRandomnessSource randomnessSource = new SimpleLayerRandomnessSource(seed);
+
+	@Inject(method = "getBiomeForNoiseGen", at = @At("RETURN"), cancellable = true)
+	private void getWeightedEndBiome(int biomeX, int biomeY, int biomeZ, CallbackInfoReturnable<Biome> cir) {
+		Biome vanillaBiome = cir.getReturnValue();
+
+		// Since all vanilla biomes are added to the registry, this will never fail.
+		// If not all vanilla biomes are added to the registry, this code will fail
+		// either way.
+		RegistryKey<Biome> vanillaKey = biomeRegistry.getKey(vanillaBiome).get();
+		RegistryKey<Biome> replacementKey;
+
+		if (vanillaKey == BiomeKeys.END_MIDLANDS || vanillaKey == BiomeKeys.END_BARRENS) {
+			// Since the pickers are statically populated by InternalBiomeData, picker will never be null.
+			WeightedBiomePicker highlandsPicker = InternalBiomeData.getEndBiomesMap().get(BiomeKeys.END_HIGHLANDS);
+			RegistryKey<Biome> highlandsKey = highlandsPicker.pickFromNoise(randomnessSource, biomeX/64.0, 0, biomeZ/64.0);
+
+			if (vanillaKey == BiomeKeys.END_MIDLANDS) {
+				// Since the pickers are statically populated by InternalBiomeData, picker will never be null.
+				WeightedBiomePicker midlandsPicker = InternalBiomeData.getEndMidlandsMap().get(highlandsKey);
+				replacementKey = midlandsPicker.pickFromNoise(randomnessSource, biomeX/64.0, 0, biomeZ/64.0);
+			} else {
+				// Since the pickers are statically populated by InternalBiomeData, picker will never be null.
+				WeightedBiomePicker barrensPicker = InternalBiomeData.getEndBarrensMap().get(highlandsKey);
+				replacementKey = barrensPicker.pickFromNoise(randomnessSource, biomeX/64.0, 0, biomeZ/64.0);
+			}
+		} else {
+			// Since the pickers are statically populated by InternalBiomeData, picker will never be null.
+			WeightedBiomePicker picker = InternalBiomeData.getEndBiomesMap().get(vanillaKey);
+			replacementKey = picker.pickFromNoise(randomnessSource, biomeX/64.0, 0, biomeZ/64.0);
```

Comment:
One thing that was never actually addressed was whether or not this stretch value of 64 was good for biome sampling. Since `TheEndBiomeSource` is very simplistic, there isn't actually anything to base this value off of and was something I came up with. Perhaps 128 or 256 would be better?

Reply:
I did some testing and it's very difficult to find a custom biome when the stretch is 128, so I'm just going to leave it as is.

Reply:
It would be nice if you throw a comment there explaining why 64 is the best value for say